core/src/main/java/org/neo4j/graphalgo/core/heavyweight/AdjacencyMatrix.java

```diff
@@ -0,0 +1,311 @@
+package org.neo4j.graphalgo.core.heavyweight;
+
+import org.apache.lucene.util.ArrayUtil;
+import org.neo4j.graphalgo.api.EdgeConsumer;
+import org.neo4j.graphalgo.api.EdgeCursor;
+import org.neo4j.graphalgo.api.WeightedEdgeConsumer;
+import org.neo4j.graphalgo.api.WeightedEdgeCursor;
+import org.neo4j.graphalgo.core.WeightMapping;
+import org.neo4j.graphdb.Direction;
+
+import java.util.Arrays;
+import java.util.Iterator;
+
+/**
+ * Relation Container built of multiple arrays. The node capacity must be constant and the node IDs have to be
+ * smaller then the capacity. The number of relations per node is limited only to the maximum array size of the VM
+ * and connections can be added dynamically.
+ *
+ * @author mknblch
+ */
+class AdjacencyMatrix {
+
+    private static final int[] EMPTY_INTS = new int[0];
+    private static final long[] EMPTY_LONGS = new long[0];
+
+    /**
+     * mapping from nodeId to outgoing degree
+     */
+    final int[] outOffsets;
+    /**
+     * mapping from nodeId to incoming degree
+     */
+    final int[] inOffsets;
+    /**
+     * matrix nodeId x [outgoing edge-ids..]
+     */
+    final int[][] outgoing;
+    /**
+     * matrix nodeId x [incoming edge-ids..]
+     */
+    final int[][] incoming;
+    /**
+     * matrix nodeId x [outgoing relation-ids..]
+     */
+    final long[][] outgoingIds;
+    /**
+     * matrix nodeId x [incoming relation-ids..]
+     */
+    final long[][] incomingIds;
+
+    AdjacencyMatrix(int nodeCount) {
+        this.outOffsets = new int[nodeCount];
+        this.inOffsets = new int[nodeCount];
+        this.outgoing = new int[nodeCount][];
+        this.incoming = new int[nodeCount][];
+        this.outgoingIds = new long[nodeCount][];
+        this.incomingIds = new long[nodeCount][];
+        Arrays.fill(outgoing, EMPTY_INTS);
+        Arrays.fill(incoming, EMPTY_INTS);
+        Arrays.fill(outgoingIds, EMPTY_LONGS);
+        Arrays.fill(incomingIds, EMPTY_LONGS);
+    }
+
+    AdjacencyMatrix(
```

Comment:
one ctor is for loading with factory, the other one is used by the FileLoader.

Task: loading from file still necessary?